Libraries/LibGfx/ICOLoader.cpp

```diff
@@ -0,0 +1,387 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/BufferStream.h>
+#include <AK/ByteBuffer.h>
+#include <AK/LexicalPath.h>
+#include <AK/MappedFile.h>
+#include <AK/NonnullOwnPtrVector.h>
+#include <LibGfx/BMPStructures.h>
+#include <LibGfx/ICOLoader.h>
+#include <LibGfx/PNGLoader.h>
+#include <LibM/math.h>
+#include <stdio.h>
+#include <string.h>
+
+//#define ICO_DEBUG
+
+namespace Gfx {
+
+// FIXME: This is in little-endian order. Maybe need a NetworkOrdered<T> equivalent eventually.
+struct ICONDIR {
+    u16 must_be_0;
+    u16 must_be_1;
+    u16 image_count;
+};
+static_assert(sizeof(ICONDIR) == 6);
+
+struct ICONDIRENTRY {
+    u8 width;
+    u8 height;
+    u8 color_count;
+    u8 reserved_0;
+    u16 planes;
+    u16 bits_per_pixel;
+    u32 size;
+    u32 offset;
+};
+static_assert(sizeof(ICONDIRENTRY) == 16);
+
+struct ImageDescriptor {
+    u16 width;
+    u16 height;
+    size_t offset;
+    size_t size;
+    RefPtr<Gfx::Bitmap> bitmap;
+};
+
+struct ICOLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        DirectoryDecoded,
+        BitmapDecoded
+    };
+    State state { NotDecoded };
+    const u8* data { nullptr };
+    size_t data_size { 0 };
+    Vector<ImageDescriptor> images;
+    size_t largest_index;
+};
+
+RefPtr<Gfx::Bitmap> load_ico(const StringView& path)
+{
+    MappedFile mapped_file(path);
+    if (!mapped_file.is_valid())
+        return nullptr;
+    ICOImageDecoderPlugin decoder((const u8*)mapped_file.data(), mapped_file.size());
+    auto bitmap = decoder.bitmap();
+    if (bitmap)
+        bitmap->set_mmap_name(String::format("Gfx::Bitmap [%dx%d] - Decoded ICO: %s", bitmap->width(), bitmap->height(), LexicalPath::canonicalized_path(path).characters()));
+    return bitmap;
+}
+
+RefPtr<Gfx::Bitmap> load_ico_from_memory(const u8* data, size_t length)
+{
+    ICOImageDecoderPlugin decoder(data, length);
+    auto bitmap = decoder.bitmap();
+    if (bitmap)
+        bitmap->set_mmap_name(String::format("Gfx::Bitmap [%dx%d] - Decoded ICO: <memory>", bitmap->width(), bitmap->height()));
+    return bitmap;
+}
+
+static Optional<size_t> decode_ico_header(BufferStream& stream)
+{
+    ICONDIR header;
+    stream.read_raw((uint8_t*)&header, sizeof(header));
+    if (stream.handle_read_failure())
+        return {};
+
+    if (header.must_be_0 != 0 || header.must_be_1 != 1)
+        return {};
+    return { header.image_count };
+}
+
+static Optional<ImageDescriptor> decode_ico_direntry(BufferStream& stream)
+{
+    ICONDIRENTRY entry;
+    stream.read_raw((uint8_t*)&entry, sizeof(entry));
+    if (stream.handle_read_failure())
+        return {};
+
+    ImageDescriptor desc = { entry.width, entry.height, entry.offset, entry.size, nullptr };
+    if (desc.width == 0)
```

Comment:
Out of interest, which web sites have favicons with invalid dimensions?

Reply:
This is actually just how the ICO format specifies a dimension of 256, since the width/height fields are only 8 bits wide. This is see on Hacker News for example.